Mathlib/Analysis/AperiodicOrder/Delone/Basic.lean

```diff
@@ -0,0 +1,172 @@
+  of `D` are at distance at least `r`;
+* **relatively dense**: there exists `R > 0` such that every point of `X`
+  is within distance `R` of some point of `D`.
+
+Delone sets arise in discrete geometry, crystallography, quasicrystals,
+aperiodic order, and tiling theory.
+
+## Main definitions
+
+* `UniformlyDiscrete (D : Set X)`
+  Positive packing radius separating distinct points of `D`.
+
+* `RelativelyDense (D : Set X)`
+  Positive covering radius so that `R`-balls around `D` cover `X`.
+
+* `DeloneSet X`
+  A structure bundling a uniformly discrete and relatively dense set.
+
+## Basic properties
+
+* `DeloneSet.nonempty`
+* `DeloneSet.exists_covering_radius`
+* `DeloneSet.exists_packing_radius`
+* `DeloneSet.dist_pos_of_ne`
+* `DeloneSet.subset_ball_singleton`
+* `DeloneSet.map`
+-/
+
+@[expose] public section
+
+open Metric
+
+variable {X Y : Type*} [MetricSpace X] [MetricSpace Y]
+
+namespace Metric
+
+/-- A set `D` in a metric space is *uniformly discrete* if there exists `r > 0`
+such that distinct points of `D` are at least distance `r` apart. -/
+def UniformlyDiscrete (D : Set X) : Prop :=
+    ∃ r > 0, ∀ ⦃x y⦄, x ∈ D → y ∈ D → x ≠ y → r ≤ dist x y
+
+/-- A set `D` in a metric space is *relatively dense* if there exists `R > 0`
+such that every point of the space is within distance `R` of some point in `D`. -/
+def RelativelyDense (D : Set X) : Prop :=
+    ∃ R > 0, ∀ x : X, ∃ y ∈ D, dist x y ≤ R
+
+/-- If `D ⊆ E` and `E` is uniformly discrete, then so is `D`. -/
+lemma uniformlyDiscrete_mono {D E : Set X} (hDE : D ⊆ E) :
+    UniformlyDiscrete E → UniformlyDiscrete D := by
+  rintro ⟨r, hr, hsep⟩
+  refine ⟨r, hr, fun x y hx hy hne ↦ ?_⟩
+  exact hsep (hDE hx) (hDE hy) hne
+
+/-- If `D ⊆ E` and `D` is relatively dense, then so is `E`. -/
+lemma relativelyDense_mono {D E : Set X} (hDE : D ⊆ E) :
+    RelativelyDense D → RelativelyDense E := by
+  rintro ⟨R, hR, hcov⟩
+  refine ⟨R, hR, fun x ↦ ?_⟩
+  rcases hcov x with ⟨y, hyD, hxy⟩
+  exact ⟨y, hDE hyD, hxy⟩
+
+end Metric
+
+namespace Delone
+
+/-- A **Delone set** in a metric space: uniformly discrete and relatively dense. -/
+structure DeloneSet (X : Type*) [MetricSpace X] where
+  /-- The underlying set of a Delone set. -/
+  (carrier : Set X)
+  /-- Uniform discreteness: distinct points of the set are separated by a positive distance. -/
+  (uniformlyDiscrete : UniformlyDiscrete carrier)
+  /-- Relative denseness: every point of the space is within some bounded distance of the set. -/
+  (relativelyDense : RelativelyDense carrier)
+
+attribute [simp] DeloneSet.carrier
+
+namespace DeloneSet
+
+/-- A Delone set is nonempty. -/
+lemma nonempty [Inhabited X] (D : DeloneSet X) : D.carrier.Nonempty := by
+  obtain ⟨R, hR, hcov⟩ := D.relativelyDense
+  rcases hcov (default : X) with ⟨y, hyD, _⟩
+  exact ⟨y, hyD⟩
+
+/-- Extract the covering radius of a Delone set. -/
+lemma exists_covering_radius (D : DeloneSet X) :
+    ∃ R > 0, ∀ x : X, ∃ y ∈ D.carrier, dist x y ≤ R :=
```

Comment:
I think I would probably instead make a `def` called `covering_radius` which picks one arbitrarily with `Classical.choose`. It makes it easier to work with in the long run.